src/rp2040/hardware_structs/include/hardware/structs/dma.h

```diff
@@ -8,46 +10,173 @@
+
+    _REG_(DMA_INTR_OFFSET)
+    // Interrupt Status (raw)
+    // 0x0000ffff [0-15]  : INTR (0): Raw interrupt status for DMA Channels 0
     io_ro_32 intr;
+
+    _REG_(DMA_INTE0_OFFSET)
+    // Interrupt Enables for IRQ 0
+    // 0x0000ffff [0-15]  : INTE0 (0): Set bit n to pass interrupts from channel n to DMA IRQ 0
     io_rw_32 inte0;
+
+    _REG_(DMA_INTF0_OFFSET)
+    // Force Interrupts
+    // 0x0000ffff [0-15]  : INTF0 (0): Write 1s to force the corresponding bits in INTE0
     io_rw_32 intf0;
+
+    _REG_(DMA_INTS0_OFFSET)
+    // Interrupt Status for IRQ 0
+    // 0x0000ffff [0-15]  : INTS0 (0): Indicates active channel interrupt requests which are currently causing IRQ 0 to be asserted
     io_rw_32 ints0;
-    uint32_t _pad1[1];
+
+    uint32_t _pad1;
+
+    _REG_(DMA_INTE1_OFFSET)
+    // Interrupt Enables for IRQ 1
+    // 0x0000ffff [0-15]  : INTE1 (0): Set bit n to pass interrupts from channel n to DMA IRQ 1
     io_rw_32 inte1;
+
+    _REG_(DMA_INTF1_OFFSET)
+    // Force Interrupts for IRQ 1
+    // 0x0000ffff [0-15]  : INTF1 (0): Write 1s to force the corresponding bits in INTE0
     io_rw_32 intf1;
+
+    _REG_(DMA_INTS1_OFFSET)
+    // Interrupt Status (masked) for IRQ 1
+    // 0x0000ffff [0-15]  : INTS1 (0): Indicates active channel interrupt requests which are currently causing IRQ 1 to be asserted
     io_rw_32 ints1;
+
+    _REG_(DMA_TIMER0_OFFSET)
     io_rw_32 timer[4];
-    io_wo_32 multi_channel_trigger;
+
+    _REG_(DMA_MULTI_CHAN_TRIGGER_OFFSET)
+    // Trigger one or more channels simultaneously
+    // 0x0000ffff [0-15]  : MULTI_CHAN_TRIGGER (0): Each bit in this register corresponds to a DMA channel
+    io_rw_32 multi_channel_trigger;
+
+    _REG_(DMA_SNIFF_CTRL_OFFSET)
+    // Sniffer Control
+    // 0x00000800 [11]    : OUT_INV (0): If set, the result appears inverted (bitwise complement) when read
+    // 0x00000400 [10]    : OUT_REV (0): If set, the result appears bit-reversed when read
+    // 0x00000200 [9]     : BSWAP (0): Locally perform a byte reverse on the sniffed data, before feeding into checksum
+    // 0x000001e0 [5-8]   : CALC (0)
+    // 0x0000001e [1-4]   : DMACH (0): DMA channel for Sniffer to observe
+    // 0x00000001 [0]     : EN (0): Enable sniffer
     io_rw_32 sniff_ctrl;
+
+    _REG_(DMA_SNIFF_DATA_OFFSET)
+    // Data accumulator for sniff hardware
     io_rw_32 sniff_data;
-    uint32_t _pad2[1];
+
+    uint32_t _pad2;
+
+    _REG_(DMA_FIFO_LEVELS_OFFSET)
+    // Debug RAF, WAF, TDF levels
+    // 0x00ff0000 [16-23] : RAF_LVL (0): Current Read-Address-FIFO fill level
+    // 0x0000ff00 [8-15]  : WAF_LVL (0): Current Write-Address-FIFO fill level
+    // 0x000000ff [0-7]   : TDF_LVL (0): Current Transfer-Data-FIFO fill level
     io_ro_32 fifo_levels;
-    io_wo_32 abort;
+
+    _REG_(DMA_CHAN_ABORT_OFFSET)
+    // Abort an in-progress transfer sequence on one or more channels
+    // 0x0000ffff [0-15]  : CHAN_ABORT (0): Each bit corresponds to a channel
+    io_rw_32 abort;
```

Comment:
Hmmm, in contrast to what I wrote above (for `MULTI_CHAN_TRIGGER`, this `CHAN_ABORT` probably **should** be `io_rw_32` because the datasheet says "The bit will remain high until any inflight transfers have been flushed through the address and data FIFOs. After writing, this register must be polled until it returns all-zero. Until this point, it is unsafe to restart the channel.". However the datasheet also labels the field as SC, whereas Appendix A in the datasheet describes SC as "This is a single bit that is written to by the processor and then **cleared on the next clock cycle**."

Hmmm, tricky... :confused:

Reply:
yeah this looks like it it is intended to be read